spatialos-sdk/src/worker/component.rs

```diff
@@ -1,35 +1,463 @@
-use spatialos_sdk_sys::worker::{
-    Schema_ComponentData, Schema_ComponentUpdate, Worker_ComponentData, Worker_ComponentUpdate,
+use crate::worker::internal::schema;
+use spatialos_sdk_sys::worker;
+use std::borrow::Borrow;
+use std::os::raw;
+use std::rc::Rc;
+use std::{
+    alloc::{self, Layout},
+    mem, ptr,
 };
 
-use crate::worker::ComponentId;
+pub type ComponentId = u32;
 
-// TODO: Wrap Schema_ComponentData
-pub struct ComponentData {
-    pub component_id: ComponentId,
-    pub schema_type: *mut Schema_ComponentData,
+pub trait ComponentUpdate<C: Component> {
+    fn merge(&mut self, update: Self);
 }
 
-impl From<&Worker_ComponentData> for ComponentData {
-    fn from(data: &Worker_ComponentData) -> Self {
-        ComponentData {
-            component_id: data.component_id,
-            schema_type: data.schema_type,
+pub trait ComponentData<C: Component> {
+    fn merge(&mut self, update: C::Update);
+}
+
+// A trait that's implemented by a type to convert to/from schema objects.
+pub trait TypeConversion
+where
+    Self: std::marker::Sized,
+{
+    fn from_type(input: &schema::SchemaObject) -> Result<Self, String>;
+    fn to_type(input: &Self, output: &mut schema::SchemaObject) -> Result<(), String>;
+}
+
+// A trait that's implemented by a component to convert to/from schema handle types.
+pub trait Component
+where
+    Self: std::marker::Sized,
+{
+    type Update;
+    type CommandRequest;
+    type CommandResponse;
+
+    fn component_id() -> ComponentId;
```

Comment:
Would this make more sense as an associated const? i.e. `const ID: ComponentId;`

Reply:
Yes, definitely. I didn't know consts in traits were also a thing.